libs/platform/ebpf_pinning_table.c

```diff
@@ -193,3 +187,146 @@ ebpf_pinning_table_delete(ebpf_pinning_table_t* pinning_table, const ebpf_utf8_s
+    if (local_pinning_entries == NULL) {
+        result = EBPF_NO_MEMORY;
+        goto Exit;
+    }
+
+    ebpf_lock_lock(&pinning_table->lock, &state);
+
+    // Loop through the entries in the hashtable.
+    next_object_name = NULL;
+    for (;;) {
+        ebpf_pinning_entry_t** next_pinning_entry = NULL;
+
+        // Find next pinning entry, if any.
+        result = ebpf_hash_table_next_key_and_value(
+            pinning_table->hash_table,
+            (const uint8_t*)((next_object_name == NULL) ? NULL : &next_object_name),
+            (uint8_t*)&next_object_name,
+            (uint8_t**)&next_pinning_entry);
+
+        if (result == EBPF_ERROR_NO_MORE_KEYS) {
+            // Reached end of hashtable.
+            result = EBPF_SUCCESS;
+            break;
+        }
+
+        if (result != EBPF_SUCCESS)
+            goto Exit_Locked;
+
+        // Skip entries that don't match the input object type.
+        if (object_type != ebpf_object_get_type((*next_pinning_entry)->object)) {
+            continue;
+        }
+
+        local_entry_count++;
+
+        if (local_entry_count > entries_array_length) {
+            ebpf_pinning_entry_t* temp_array;
+            uint16_t old_length = entries_array_length;
+            uint16_t index;
+
+            // Allocate a bigger output array with double the estimated size.
+            entries_array_length <<= 1;
+            if (entries_array_length < old_length) {
+                // Overflow occured.
+                result = EBPF_ERROR_ARITHMETIC_OVERFLOW;
+                goto Exit_Locked;
+            }
+            temp_array = (ebpf_pinning_entry_t*)ebpf_allocate(
+                sizeof(ebpf_pinning_entry_t) * entries_array_length, EBPF_MEMORY_NO_EXECUTE);
+            if (temp_array == NULL) {
+                result = EBPF_NO_MEMORY;
+                goto Exit_Locked;
+            }
+
+            // Copy over the old array.
+            for (index = 0; index < (local_entry_count - 1); index++) {
+                ebpf_pinning_entry_t* source = &local_pinning_entries[index];
+                ebpf_pinning_entry_t* destination = &temp_array[index];
+
+                *destination = *source;
+            }
+
+            // Free old array.
+            ebpf_free(&local_pinning_entries);
+
+            // Set array to new buffer.
+            local_pinning_entries = temp_array;
+        }
+
+        // Copy the next pinning entry to a new entry in the output array.
+        new_entry = &local_pinning_entries[local_entry_count - 1];
+        new_entry->object = (*next_pinning_entry)->object;
+
+        // Take reference on underlying ebpf_object.
+        ebpf_object_acquire_reference(new_entry->object);
+
+        // Duplicate pinning object name.
+        result = ebpf_duplicate_utf8_string(&new_entry->name, &(*next_pinning_entry)->name);
+        if (result != EBPF_SUCCESS)
+            goto Exit_Locked;
+    }
+
+Exit_Locked:
```

Comment:
Prefer 1 exit label, with flags to control what cleanup is needed.

Reply:
Done.